fdbserver/RestoreApplier.actor.cpp

```diff
@@ -151,85 +153,204 @@ ACTOR Future<Void> applyToDB(Reference<RestoreApplierData> self, Database cx) {
+	// uncommittedTxnId: The id of the most recent succeeded txn. Used to recover the failed txn id in retry
+	// lastTxnHasError: Does the last txn has error. TODO: Only need to handle txn_commit_unknown error
+	state Version curTxnId = 0; 
+	state Version uncommittedTxnId = 0; 
+	state bool lastTxnHasError = false; 
+
+	// Decide when to commit a transaction. We buffer enough mutations in a txn before commit the txn
+	state bool startNextVersion = false; // The next txn will include mutations in next version
+	state int numAtomicOps = 0;
 	state double transactionSize = 0;
-	loop {
+
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+
+	loop { // Transaction retry loop
 		try {
-			tr->reset();
-			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-			transactionSize = 0;
+			// Check if the transaction succeeds
+			if (lastTxnHasError) {
+				tr->reset();
+				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+				Optional<Value> txnSucceeded = wait(tr->get(restoreApplierKeyFor(self->id(), curTxnId)));
+				if (!txnSucceeded.present()) {
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusFailed", curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("CurrentFailedTxnId", curIndexInCurTxn)
+					    .detail("UncommittedTxnId", uncommittedTxnId)
+					    .detail("CurIteratorVersion", curItInCurTxn->first)
+					    .detail("StartIteratorVersionInUncommittedTxn", startItInUncommittedTxn->first)
+					    .detail("CurrentIndexInFailedTxn", curIndexInCurTxn)
+					    .detail("StartIndexInUncommittedTxn", startIndexInUncommittedTxn)
+					    .detail("NumIncludedAtomicOps", numAtomicOps);
+					// Re-execute uncommitted txn
+					curItInCurTxn = startItInUncommittedTxn;
+					curIndexInCurTxn = startIndexInUncommittedTxn;
+					curTxnId = uncommittedTxnId;
+
+					numAtomicOps = 0;
+					transactionSize = 0;
+					startNextVersion = false;
+
+					lastTxnHasError = false;
+					continue;
+				} else {
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusSucceeded", curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("CurrentSucceedTxnId", curIndexInCurTxn)
+					    .detail("CurIteratorVersion", curItInCurTxn->first)
+					    .detail("CurrentIteratorMutations", curItInCurTxn->second.size())
+					    .detail("CurrentIndexInSucceedTxn", curIndexInCurTxn)
+					    .detail("NumIncludedAtomicOps", numAtomicOps);
 
-			for (; it != self->kvOps.end(); ++it) {
-				numVersion++;
-				//TraceEvent("FastRestore").detail("Applier", self->id()).detail("ApplyKVsToDBVersion", it->first);
-				state MutationRef m;
-				for (; index < it->second.size(); ++index) {
-					m = it->second[index];
+					// Skip else, and execute the logic when a txn succeed
+				}
```

Comment:
I feel this is dead code. If `lastTxnHasError` is true, then `txnSucceeded` is always empty.

Reply:
This is not dead code. It is a must have code for transactions with atomic operations.

Transaction may throw the `commit_unknown_result` exception, which means the transaction may or may not succeed. The whole point of using `Optional<Value> txnSucceeded = wait(tr->get(restoreApplierKeyFor(self->id(), curTxnId)));` is to tell if the transaction succeeded when `commit_unknown_result ` happens.